### PR TITLE
feature: mixed sorting envoyfilter

### LINKED ISSUE #3371: Enable Mixed Sorting Between EnvoyFilter and WasmPlugin
## Problem Statement

Currently, Higress is increasingly using Golang filters deployed through Istio EnvoyFilter CRD. However, there's a critical limitation: EnvoyFilter sorting logic and WasmPlugin sorting logic are completely isolated, making it impossible to achieve mixed ordering between them. For example, placing a rag-mcp Golang filter after jwt-auth plugin cannot be precisely controlled.

## Proposed Solution

Extend the EnvoyFilter CRD with `wasmPhase` and `wasmPriority` fields to enable mixed sorting with WasmPlugins while maintaining full backward compatibility.

### 1. CRD Extension

Add new fields to EnvoyFilter spec in `manifests/charts/base/crds/crd-all.gen.yaml` [1](#0-0) :

```yaml
wasmPhase:
  description: Determines where in the filter chain this EnvoyFilter should be injected for mixed sorting with WasmPlugins
  enum: [UNSPECIFIED_PHASE, AUTHN, AUTHZ, STATS]
  type: string
wasmPriority:
  description: Determines ordering of EnvoyFilter in the same wasmPhase for mixed sorting with WasmPlugins
  type: integer
```

### 2. Core Sorting Logic Enhancement

Modify the `initEnvoyFilters` function in `pilot/pkg/model/push_context.go` [2](#0-1)  to implement mixed sorting:

```go
sort.Slice(envoyFilterConfigs, func(i, j int) bool {
    ifilter := envoyFilterConfigs[i].Spec.(*networking.EnvoyFilter)
    jfilter := envoyFilterConfigs[j].Spec.(*networking.EnvoyFilter)
    
    // Check if wasmPhase is set
    iHasWasmPhase := ifilter.WasmPhase != extensions.PluginPhase_UNSPECIFIED_PHASE && ifilter.WasmPhase != 0
    jHasWasmPhase := jfilter.WasmPhase != extensions.PluginPhase_UNSPECIFIED_PHASE && jfilter.WasmPhase != 0
    
    // Both have wasmPhase: use mixed sorting
    if iHasWasmPhase && jHasWasmPhase {
        if ifilter.WasmPhase != jfilter.WasmPhase {
            return phaseOrder(ifilter.WasmPhase) < phaseOrder(jfilter.WasmPhase)
        }
        if ifilter.WasmPriority != jfilter.WasmPriority {
            return ifilter.WasmPriority > jfilter.WasmPriority
        }
    }
    
    // Only one has wasmPhase: prioritize the one with wasmPhase
    if iHasWasmPhase && !jHasWasmPhase {
        return true
    }
    if !iHasWasmPhase && jHasWasmPhase {
        return false
    }
    
    // Neither has wasmPhase: use original sorting logic
    if ifilter.Priority != jfilter.Priority {
        return ifilter.Priority < jfilter.Priority
    }
    // ... rest of original logic
})
```

### 3. LDS Generation Integration

Update the LDS generation flow in `pilot/pkg/xds/lds.go` [3](#0-2)  to apply mixed sorting before generating listeners.

### 4. Filter Chain Mixed Sorting

Implement mixed sorting in the HTTP filter chain, integrating with existing WasmPlugin handling from `pilot/pkg/networking/core/v1alpha3/extension/wasmplugin.go` [4](#0-3) .

## Usage Example

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: custom-authn-filter
spec:
  wasmPhase: AUTHN          # Enable mixed sorting
  wasmPriority: 100         # Mixed sorting priority
  priority: 10              # Traditional priority (backward compatible)
  configPatches:
    - applyTo: HTTP_FILTER
      match:
        context: SIDECAR_INBOUND
      patch:
        operation: INSERT_BEFORE
        value:
          name: envoy.filters.http.custom_authn
          typed_config:
            "@type": type.googleapis.com/envoy.extensions.filters.http.custom_authn.v3.CustomAuthn
```

## Benefits

1. **Full Backward Compatibility**: EnvoyFilters without new fields maintain original behavior
2. **Flexible Control**: Selective enablement of mixed sorting via new fields
3. **Unified Ordering**: True mixed sorting between EnvoyFilter and WasmPlugin
4. **Performance Optimized**: Unified processing before LDS generation

## Implementation Roadmap

1. **Phase 1**: Add new fields to CRD and data structures
2. **Phase 2**: Implement mixed sorting logic with feature flag
3. **Phase 3**: Comprehensive testing and documentation
4. **Phase 4**: Enable by default

## Call for Contributors

We're seeking community contributions to implement this feature. Key areas needing work:

- CRD schema updates and validation
- PushContext sorting algorithm implementation
- LDS generation integration
- Comprehensive test coverage
- Documentation and examples

## Labels

`enhancement` `envoy-filter` `wasm-plugin` `sorting` `mixed-ordering`

---

**Note**: This implementation maintains complete backward compatibility while providing the flexibility needed for mixed filter ordering scenarios in Higress deployments.

## Notes

The proposed solution leverages existing infrastructure in the Istio control plane, specifically the PushContext's configuration processing capabilities and the LDS generation pipeline. The mixed sorting logic integrates seamlessly with current WasmPlugin handling through the `PopAppend` mechanism already in place.

Wiki pages you might want to explore:
- [Control Plane (istiod) (higress-group/istio)](/wiki/higress-group/istio#2)
- [Configuration Processing (higress-group/istio)](/wiki/higress-group/istio#2.2)

### Citations

**File:** manifests/charts/base/crds/crd-all.gen.yaml (L3021-3030)
```yaml
          spec:
            description: 'Customizing Envoy configuration generated by Istio. See
              more details at: https://istio.io/docs/reference/config/networking/envoy-filter.html'
            properties:
              configPatches:
                description: One or more patches with match conditions.
                items:
                  properties:
                    applyTo:
                      enum:
```

**File:** pilot/pkg/model/push_context.go (L2017-2033)
```go
	sort.Slice(envoyFilterConfigs, func(i, j int) bool {
		ifilter := envoyFilterConfigs[i].Spec.(*networking.EnvoyFilter)
		jfilter := envoyFilterConfigs[j].Spec.(*networking.EnvoyFilter)
		if ifilter.Priority != jfilter.Priority {
			return ifilter.Priority < jfilter.Priority
		}
		// If priority is same fallback to name and creation timestamp, else use priority.
		// If creation time is the same, then behavior is nondeterministic. In this case, we can
		// pick an arbitrary but consistent ordering based on name and namespace, which is unique.
		// CreationTimestamp is stored in seconds, so this is not uncommon.
		if envoyFilterConfigs[i].CreationTimestamp != envoyFilterConfigs[j].CreationTimestamp {
			return envoyFilterConfigs[i].CreationTimestamp.Before(envoyFilterConfigs[j].CreationTimestamp)
		}
		in := envoyFilterConfigs[i].Name + "." + envoyFilterConfigs[i].Namespace
		jn := envoyFilterConfigs[j].Name + "." + envoyFilterConfigs[j].Namespace
		return in < jn
	})
```

**File:** pilot/pkg/networking/core/v1alpha3/extension/wasmplugin.go (L49-60)
```go
// PopAppend takes a list of filters and a set of WASM plugins, keyed by phase. It will remove all
// plugins of a provided phase from the WASM plugin set and append them to the list of filters
func PopAppend(list []*hcm.HttpFilter,
	filterMap map[extensions.PluginPhase][]*model.WasmPluginWrapper,
	phase extensions.PluginPhase,
) []*hcm.HttpFilter {
	for _, ext := range filterMap[phase] {
		list = append(list, toEnvoyHTTPFilter(ext))
	}
	delete(filterMap, phase)
	return list
}
```

---
## Problem Statement

Currently, Higress is increasingly using Golang filters deployed through Istio EnvoyFilter CRD. However, there's a critical limitation: EnvoyFilter sorting logic and WasmPlugin sorting logic are completely isolated, making it impossible to achieve mixed ordering between them. For example, placing a rag-mcp Golang filter after jwt-auth plugin cannot be precisely controlled.

## Proposed Solution

Extend the EnvoyFilter CRD with `wasmPhase` and `wasmPriority` fields to enable mixed sorting with WasmPlugins while maintaining full backward compatibility.

### 1. CRD Extension

Add new fields to EnvoyFilter spec in `manifests/charts/base/crds/crd-all.gen.yaml` [1](#0-0) :

```yaml
wasmPhase:
  description: Determines where in the filter chain this EnvoyFilter should be injected for mixed sorting with WasmPlugins
  enum: [UNSPECIFIED_PHASE, AUTHN, AUTHZ, STATS]
  type: string
wasmPriority:
  description: Determines ordering of EnvoyFilter in the same wasmPhase for mixed sorting with WasmPlugins
  type: integer
```

### 2. Core Sorting Logic Enhancement

Modify the `initEnvoyFilters` function in `pilot/pkg/model/push_context.go` [2](#0-1)  to implement mixed sorting:

```go
sort.Slice(envoyFilterConfigs, func(i, j int) bool {
    ifilter := envoyFilterConfigs[i].Spec.(*networking.EnvoyFilter)
    jfilter := envoyFilterConfigs[j].Spec.(*networking.EnvoyFilter)
    
    // Check if wasmPhase is set
    iHasWasmPhase := ifilter.WasmPhase != extensions.PluginPhase_UNSPECIFIED_PHASE && ifilter.WasmPhase != 0
    jHasWasmPhase := jfilter.WasmPhase != extensions.PluginPhase_UNSPECIFIED_PHASE && jfilter.WasmPhase != 0
    
    // Both have wasmPhase: use mixed sorting
    if iHasWasmPhase && jHasWasmPhase {
        if ifilter.WasmPhase != jfilter.WasmPhase {
            return phaseOrder(ifilter.WasmPhase) < phaseOrder(jfilter.WasmPhase)
        }
        if ifilter.WasmPriority != jfilter.WasmPriority {
            return ifilter.WasmPriority > jfilter.WasmPriority
        }
    }
    
    // Only one has wasmPhase: prioritize the one with wasmPhase
    if iHasWasmPhase && !jHasWasmPhase {
        return true
    }
    if !iHasWasmPhase && jHasWasmPhase {
        return false
    }
    
    // Neither has wasmPhase: use original sorting logic
    if ifilter.Priority != jfilter.Priority {
        return ifilter.Priority < jfilter.Priority
    }
    // ... rest of original logic
})
```

### 3. LDS Generation Integration

Update the LDS generation flow in `pilot/pkg/xds/lds.go` [3](#0-2)  to apply mixed sorting before generating listeners.

### 4. Filter Chain Mixed Sorting

Implement mixed sorting in the HTTP filter chain, integrating with existing WasmPlugin handling from `pilot/pkg/networking/core/v1alpha3/extension/wasmplugin.go` [4](#0-3) .

## Usage Example

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: custom-authn-filter
spec:
  wasmPhase: AUTHN          # Enable mixed sorting
  wasmPriority: 100         # Mixed sorting priority
  priority: 10              # Traditional priority (backward compatible)
  configPatches:
    - applyTo: HTTP_FILTER
      match:
        context: SIDECAR_INBOUND
      patch:
        operation: INSERT_BEFORE
        value:
          name: envoy.filters.http.custom_authn
          typed_config:
            "@type": type.googleapis.com/envoy.extensions.filters.http.custom_authn.v3.CustomAuthn
```

## Benefits

1. **Full Backward Compatibility**: EnvoyFilters without new fields maintain original behavior
2. **Flexible Control**: Selective enablement of mixed sorting via new fields
3. **Unified Ordering**: True mixed sorting between EnvoyFilter and WasmPlugin
4. **Performance Optimized**: Unified processing before LDS generation

## Implementation Roadmap

1. **Phase 1**: Add new fields to CRD and data structures
2. **Phase 2**: Implement mixed sorting logic with feature flag
3. **Phase 3**: Comprehensive testing and documentation
4. **Phase 4**: Enable by default

## Call for Contributors

We're seeking community contributions to implement this feature. Key areas needing work:

- CRD schema updates and validation
- PushContext sorting algorithm implementation
- LDS generation integration
- Comprehensive test coverage
- Documentation and examples

## Labels

`enhancement` `envoy-filter` `wasm-plugin` `sorting` `mixed-ordering`

---

**Note**: This implementation maintains complete backward compatibility while providing the flexibility needed for mixed filter ordering scenarios in Higress deployments.

## Notes

The proposed solution leverages existing infrastructure in the Istio control plane, specifically the PushContext's configuration processing capabilities and the LDS generation pipeline. The mixed sorting logic integrates seamlessly with current WasmPlugin handling through the `PopAppend` mechanism already in place.

Wiki pages you might want to explore:
- [Control Plane (istiod) (higress-group/istio)](/wiki/higress-group/istio#2)
- [Configuration Processing (higress-group/istio)](/wiki/higress-group/istio#2.2)

### Citations

**File:** manifests/charts/base/crds/crd-all.gen.yaml (L3021-3030)
```yaml
          spec:
            description: 'Customizing Envoy configuration generated by Istio. See
              more details at: https://istio.io/docs/reference/config/networking/envoy-filter.html'
            properties:
              configPatches:
                description: One or more patches with match conditions.
                items:
                  properties:
                    applyTo:
                      enum:
```

**File:** pilot/pkg/model/push_context.go (L2017-2033)
```go
	sort.Slice(envoyFilterConfigs, func(i, j int) bool {
		ifilter := envoyFilterConfigs[i].Spec.(*networking.EnvoyFilter)
		jfilter := envoyFilterConfigs[j].Spec.(*networking.EnvoyFilter)
		if ifilter.Priority != jfilter.Priority {
			return ifilter.Priority < jfilter.Priority
		}
		// If priority is same fallback to name and creation timestamp, else use priority.
		// If creation time is the same, then behavior is nondeterministic. In this case, we can
		// pick an arbitrary but consistent ordering based on name and namespace, which is unique.
		// CreationTimestamp is stored in seconds, so this is not uncommon.
		if envoyFilterConfigs[i].CreationTimestamp != envoyFilterConfigs[j].CreationTimestamp {
			return envoyFilterConfigs[i].CreationTimestamp.Before(envoyFilterConfigs[j].CreationTimestamp)
		}
		in := envoyFilterConfigs[i].Name + "." + envoyFilterConfigs[i].Namespace
		jn := envoyFilterConfigs[j].Name + "." + envoyFilterConfigs[j].Namespace
		return in < jn
	})
```

**File:** pilot/pkg/networking/core/v1alpha3/extension/wasmplugin.go (L49-60)
```go
// PopAppend takes a list of filters and a set of WASM plugins, keyed by phase. It will remove all
// plugins of a provided phase from the WASM plugin set and append them to the list of filters
func PopAppend(list []*hcm.HttpFilter,
	filterMap map[extensions.PluginPhase][]*model.WasmPluginWrapper,
	phase extensions.PluginPhase,
) []*hcm.HttpFilter {
	for _, ext := range filterMap[phase] {
		list = append(list, toEnvoyHTTPFilter(ext))
	}
	delete(filterMap, phase)
	return list
}
```

### PR DESCRIPTION
**Pull request by Aias00**
## Ⅰ. Describe what this PR did

support mixed sorting envoyfilter

link: https://github.com/alibaba/higress/issues/3371
link: https://github.com/higress-group/api/pull/5
link: https://github.com/higress-group/istio/pull/48

## Ⅱ. Does this pull request fix one issue?

fixes: #3371

## Ⅲ. Why don't you add test cases (unit test/integration test)? 

N/A - CRD schema-only change. The EnvoyFilter schema is generated from higress-group/api#5 and validated with client-side CRD dry-run.

## Ⅳ. Describe how to verify it

- Confirmed istio/api is at efc0fe4, the merge commit for higress-group/api#5 including 6b0cde2.
- Matched wasmPhase/wasmPriority schema text against istio/api/kubernetes/customresourcedefinitions.gen.yaml.
- Ran `git diff --check origin/main...HEAD`.
- Ran `kubectl apply --dry-run=client --validate=false -f helm/core/crds/istio-envoyfilter.yaml`.
- Ran `kubectl apply --dry-run=client --validate=false -f hgctl/pkg/manifests/istiobase/crds/crd-all.gen.yaml`.

## Ⅵ. AI Coding Tool Usage Checklist (if applicable)

**Please check all applicable items:**

- [ ] **For new standalone features** (e.g., new wasm plugin or golang-filter plugin):
  - [ ] I have created a `design/` directory in the plugin folder
  - [ ] I have added the design document to the `design/` directory
  - [ ] I have included the AI Coding summary below

- [ ] **For regular updates/changes** (not new plugins):
  - [ ] I have provided the prompts/instructions I gave to the AI Coding tool below
  - [ ] I have included the AI Coding summary below

### CODE REVIEW

**Comment by CH3CHO** [comment]
CC @EndlessSeeker

File: hgctl/pkg/manifests/istiobase/crds/crd-all.gen.yaml [hunk 1]
```diff
@@ -3209,6 +3209,20 @@ spec:
                   within a context.
                 format: int32
                 type: integer
+              wasmPhase:
```
Comment (CH3CHO) [inline hgctl/pkg/manifests/istiobase/crds/crd-all.gen.yaml:3212]: ⚠️ This file's first line is `DO NOT EDIT - Generated by Cue OpenAPI generator based on Istio APIs.` and the generator is `api/gen.sh` (`cue-gen -verbose -f=./cue.yaml -crd=true`). A hand-edit here will be silently overwritten the next time the generator runs, and any small normalization the Cue tool would apply (enum-value ordering, int32→int64 widening, description line wrapping) is now at risk of drift. Please regenerate this file via `api/gen.sh` after bumping the `istio/api` submodule to a SHA that already includes the merged `higress-group/api#5` (commit `6b0cde2`), or document in the PR description why a manual edit is needed for this release window.

The companion change in `helm/core/crds/istio-envoyfilter.yaml` is the hand-maintained chart copy and does not need regeneration.

**Review by CH3CHO — CHANGES_REQUESTED** [review]
Thanks for the CRD sync. The two `wasmPhase`/`wasmPriority` insertions look correct and the two CRD copies are byte-identical, which is good. I have one blocker and a couple of polish items.

**Blocker (Issue #1)**
`hgctl/pkg/manifests/istiobase/crds/crd-all.gen.yaml` is an auto-generated artifact — its first line says `DO NOT EDIT - Generated by Cue OpenAPI generator based on Istio APIs.` and it's produced by `api/gen.sh` (`cue-gen -verbose -f=./cue.yaml -crd=true`). Hand-editing the generated file will be silently overwritten the next time the generator runs, and any small normalization (e.g. enum order, int32→int64 widening, description line wrapping) the Cue tool would have made is now at risk of drift. Please either:

  1. Run `api/gen.sh` (with the istio submodule bumped to a SHA that already includes `higress-group/api#5`, which is merged) and commit the regenerated file so it matches the generator's expected output exactly; or
  2. If a manual edit is intentional for this release window, add a comment in the PR description explaining why and link the tracking issue for re-syncing.

The `helm/core/crds/istio-envoyfilter.yaml` change is the hand-maintained chart copy and is fine as-is.

**Polish (Issue #2)**
The PR template's `Why don't you add test cases` section is empty. For a CRD-only change the natural answer is `N/A — schema-only, validated by re-running api/gen.sh and kubectl apply --dry-server`. A one-line note would unblock this for future reviewers.

**Optional (Issues #3/#4)**
`wasmPriority` is unbounded int32 — consistent with the existing `priority` field, no action needed. A short note in the `wasmPhase` description explaining the phase order (`AUTHN` → `AUTHZ` → `STATS`) would help users pick the right value.

Also noting for the record: the two earlier Copilot review comments on this PR (the helm/core CRD sync, and the `no Go code references` concern) are both resolved — the first by the second commit `2e488006`, the second by the already-merged linked PRs `higress-group/api#5` and `higress-group/istio#48`, whose merge commits are already in the istio submodules of this branch. So once the auto-generated file is handled, this is good to merge.

**Comment by Aias00** [comment]
I’ve addressed the CRD review blocker in 6422bda.

  The PR is now updated with the latest origin/main. For the CRD change, I kept istio/api at efc0fe4, which already includes the api#5 changes, and updated both EnvoyFilter CRD copies so the wasmPhase/wasmPriority
  schema matches the generated output from istio/api/kubernetes/customresourcedefinitions.gen.yaml.

  I also filled in the PR test and verification sections for this CRD-only change.

File: hgctl/pkg/manifests/istiobase/crds/crd-all.gen.yaml (hunk 1 above)
Comment (CH3CHO) [reply hgctl/pkg/manifests/istiobase/crds/crd-all.gen.yaml:3212]: 已确认解决。`fix(crds): sync EnvoyFilter wasm sorting schema`（6422bda9）将两个 CRD 副本的 `wasmPhase` / `wasmPriority` description 同步为 `istio/api@efc0fe4:kubernetes/customresourcedefinitions.gen.yaml` 生成器输出口径，逐字符比对一致；helm 与 hgctl 两份 CRD 字节相同。polish 项（test 段 / phase 顺序注释）已分别覆盖：PR 模板 test 段已写明 `N/A - CRD schema-only change`，phase 顺序留作生成器原样以避免再次漂移。

File: hgctl/pkg/manifests/istiobase/crds/crd-all.gen.yaml [hunk 2]
```diff
@@ -3209,6 +3209,20 @@ spec:
...
+                - UNSPECIFIED_PHASE
+                - AUTHN
+                - AUTHZ
+                - STATS
+                type: string
+              wasmPriority:
+                description: Determines ordering of EnvoyFilter in the same wasmPhase
+                  for mixed sorting with WasmPlugins.
+                format: int32
+                type: integer
```
[the remark this answers is not in the document]
Comment (CH3CHO) [reply hgctl/pkg/manifests/istiobase/crds/crd-all.gen.yaml:3226]: 明确一下：本仓库在 mixed sorting 链路中的职责是 CRD/installer 增量，控制面实现在上游子模块。`istio/api@efc0fe4:networking/v1alpha3/envoy_filter.proto` 已含 `wasm_phase = 7` / `wasm_priority = 8`（由 `higress-group/api#5` 提交 6b0cde2 合并而来），控制面排序逻辑在 `higress-group/istio#48`（已合入 `istio/istio` 子模块）。本 PR 仅同步 CRD schema，让 k8s API server 接受这些字段。已逐字符核对 `istio/api@efc0fe4:kubernetes/customresourcedefinitions.gen.yaml` EnvoyFilter 块与本 PR 的 `wasmPhase` / `wasmPriority` 完全一致。

**Review by CH3CHO — APPROVED** [review]
CRD schema 同步完整，覆盖了之前的 blocker。两条 review thread 均已回复并 resolve。

复核要点：
- `wasmPhase` / `wasmPriority` description 已与 `istio/api@efc0fe4:kubernetes/customresourcedefinitions.gen.yaml` 逐字符一致。
- `helm/core/crds/istio-envoyfilter.yaml` 与 `hgctl/pkg/manifests/istiobase/crds/crd-all.gen.yaml` 在 EnvoyFilter 块上字节相同，helm/hgctl 两条安装路径无漂移。
- 子模块 `istio/api@efc0fe4` / `istio/istio@de2c9628` 已包含上游 `higress-group/api#5` (6b0cde2) 与 `higress-group/istio#48` 的合并，本仓 CRD 仅是 schema 增量。
- 子模块未 bump，正确。
- 全部 CI（lint / build / coverage / conformance / wasmplugin / DCO / CLA / FOSSA / sensitive-scan）通过。

非阻塞观察（不阻塞合并）：`hgctl/pkg/manifests/istiobase/crds/crd-all.gen.yaml` 的 `DO NOT EDIT` header 与现状不符——当前没有工作流会自动重生它，全靠人工同步。可后续单独提一个 PR 去掉 header 或补一个 sync 脚本。

LGTM ✅
